### PR TITLE
[FEATURE][FIX] Add MiddlewareUtil providing convenience functions to all middleware

### PR DESCRIPTION
**Pull request by RandomByte**
In order to correctly handle requests with URIs containing escape
sequences like "%20" (=whitespace), middleware implementations need to
decode any escape sequences in incoming request. To make this easier a
new MiddlewareUtil class is added. Every MiddlewareManager creates one
instance of this class and provides it to every standard middleware.

Custom middleware receive an interface to this instance with a set of
methods restricted by the specification version of the custom middleware
extension. This is yet to be implemented in a follow-up change as the
middleware repository does not yet provide specification version
information.

### CODE REVIEW

File: test/lib/server/middleware/MiddlewareUtil.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+const sinon = require("sinon");
+const mock = require("mock-require");
+const MiddlewareUtil = require("../../../../lib/middleware/MiddlewareUtil");
+
+test.afterEach.always((t) => {
+	sinon.restore();
+	mock.stopAll();
+});
+
+test("getPathname", async (t) => {
```
Comment (matz3) [inline test/lib/server/middleware/MiddlewareUtil.js]: Should be "test.serial" as it mocks "parseurl"

File: lib/middleware/MiddlewareUtil.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		};
+		switch (specVersion) {
+		case "0.1":
+		case "1.0":
+		case "1.1":
+			return undefined;
+		case "2.0":
+			return baseInterface;
+		default:
+			throw new Error(`MiddlewareUtil: Unkown or unsupported specification version ${specVersion}`);
```
Comment (matz3) [inline lib/middleware/MiddlewareUtil.js]:
```suggestion
			throw new Error(`MiddlewareUtil: Unknown or unsupported specification version ${specVersion}`);
```

File: lib/middleware/MiddlewareUtil.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+ * @public
+ * @memberof module:@ui5/server.middleware
+ */
+class MiddlewareUtil {
+	/**
+	 * Get an interface to an instance of this class that only provides those functions
+	 * that are supported by the given custom middleware extension specification version.
+	 *
+	 * @param {string} specVersion Specification Version of custom middleware extension
+	 * @returns {object} An object with bound instance methods supported by the given specificationv ersion
```
Comment (matz3) [inline lib/middleware/MiddlewareUtil.js]:
```suggestion
	 * @returns {object} An object with bound instance methods supported by the given specification version
```

File: lib/middleware/MiddlewareUtil.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	/**
+	 * Returns the [pathname]{@link https://developer.mozilla.org/en-US/docs/Web/API/URL/pathname}
+	 * of a given request. Any escape sequences will be decoded.
+	 * </br></br>
+	 * This method is only available to custom middleware extensions defining
+	 * <b>specification version 2.0 and above</b>.
+	 *
+	 * @param {object} req Express request object
```
Comment (matz3) [inline lib/middleware/MiddlewareUtil.js]: I think we should not document Express specifics here. We should only expect a standard request object aka Incoming Message: https://nodejs.org/api/http.html#http_class_http_incomingmessage

Just removing Express should be sufficient IMO.

File: test/lib/server/middleware/MiddlewareUtil.js (hunk 1 above)
Comment (RandomByte) [reply test/lib/server/middleware/MiddlewareUtil.js]: Done

File: lib/middleware/MiddlewareUtil.js (hunk 4 above)
Comment (RandomByte) [reply lib/middleware/MiddlewareUtil.js]: Done

**Comment by matz3** [comment]
Will `[FEATURE][FIX]` create entries for both changelogs?
Otherwise you might squash the changes into two commits so that both are part of the changelog.

**Pull request merged**